base/sort.jl

```diff
@@ -528,75 +532,92 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, ::InsertionSortAlg,
     return v
 end
 
-# selectpivot!
+# select a pivot for QuickSort
 #
-# Given 3 locations in an array (lo, mi, and hi), sort v[lo], v[mi], v[hi]) and
-# choose the middle value as a pivot
-#
-# Upon return, the pivot is in v[lo], and v[hi] is guaranteed to be
-# greater than the pivot
+# This method is redefined to rand(lo:hi) in Random.jl
+# We can't use rand here because it is not available in Core.Compiler and
+# because rand is defined in the stdlib Random.jl after sorting is used in Base.
+select_pivot(lo::Integer, hi::Integer) = midpoint(lo, hi)
```

Comment:
@timholy will this cause an invalidation problem?

Reply:
also, does the compiler need a slightly better algorithm (e.g. median of 3)? I'm not sure that the compiler won't hit the edge cases that would turn this quadratic (and we've been spending a lot of effort removing quadratic compiler problems).

Reply:
All these shenanigans should be unnecessary after https://github.com/JuliaLang/julia/pull/46679 because we will then be able to simply depend on Random.

Reply:
#47066 alone (not needing all of #46679) should make "does the compiler need a slightly better algorithm (e.g. median of 3)?" outdated.

Reply:
As for invalidations, I'm no Tim Holy, but I injected the following snippet into Random.jl right before the method redefinition
```julia
let
    function backedges(f::Function)
        s = Set{Base.MethodInstance}()
        for m in methods(f)
            for mi in m.specializations
                if mi !== nothing
                    backedges!(s, mi)
                end
            end
        end
        s
    end

    function backedges!(s::Set{Base.MethodInstance}, m::Base.MethodInstance)
        if isdefined(m, :backedges)
            for mi in m.backedges
                if mi ∉ s
                    push!(s, mi)
                    backedges!(s, mi)
                end
            end
        end
        s
    end

    @show length(backedges(Base.Sort.select_pivot))
end
```
and got the result
```
length(backedges(Base.Sort.select_pivot)) = 213
```
So there are 213 invalidations at that time. Luckely, Random is loaded before precompilation statements are executed. With and without the redefinition in Random.jl I got about* 3670 direct and indirect backedges from `Base.Sort.select_pivot` which makes me think that the invalidated methodinstances get recompiled later on in the build process so there should be no impact on latency after the build finishes.

*there is a bit of variability because of my manual interaction with the REPL scrollback buffer which causes various methods to be compiled)

I think this result resolves this conversation.